sentry-android-replay/src/main/java/io/sentry/android/replay/capture/SessionCaptureStrategy.kt

```diff
@@ -0,0 +1,144 @@
+        // tagged with the replay that might never be sent when we're recording in buffer mode
+        hub?.configureScope { it.replayId = currentReplayId.get() }
+    }
+
+    override fun pause() {
+        createCurrentSegment("pause") { segment ->
+            if (segment is ReplaySegment.Created) {
+                segment.capture(hub)
+
+                currentSegment.getAndIncrement()
+            }
+        }
+        super.pause()
+    }
+
+    override fun stop() {
+        val replayCacheDir = cache?.replayCacheDir
+        createCurrentSegment("stop") { segment ->
+            if (segment is ReplaySegment.Created) {
+                segment.capture(hub)
+            }
+            FileUtils.deleteRecursively(replayCacheDir)
+        }
+        hub?.configureScope { it.replayId = SentryId.EMPTY_ID }
+        super.stop()
+    }
+
+    override fun sendReplayForEvent(event: SentryEvent, hint: Hint, onSegmentSent: () -> Unit) {
+        // don't ask me why
+        event.setTag("replayId", currentReplayId.get().toString())
+        if (!event.isCrashed) {
+            options.logger.log(DEBUG, "Replay is already running in 'session' mode, not capturing for event %s", event.eventId)
+        } else {
+            options.logger.log(DEBUG, "Replay is already running in 'session' mode, capturing last segment for crashed event %s", event.eventId)
+            createCurrentSegment("send_replay_for_event") { segment ->
+                if (segment is ReplaySegment.Created) {
+                    segment.capture(hub, hint)
+                }
+            }
+        }
+    }
+
+    override fun onScreenshotRecorded(bitmap: Bitmap) {
+        // have to do it before submitting, otherwise if the queue is busy, the timestamp won't be
+        // reflecting the exact time of when it was captured
+        val frameTimestamp = dateProvider.currentTimeMillis
+        val height = recorderConfig.recordingHeight
+        val width = recorderConfig.recordingWidth
+        replayExecutor.submitSafely(options, "$TAG.add_frame") {
+            cache?.addFrame(bitmap, frameTimestamp)
+
+            val now = dateProvider.currentTimeMillis
+            if ((now - segmentTimestamp.get().time >= options.experimental.sessionReplay.sessionSegmentDuration)) {
+                val currentSegmentTimestamp = segmentTimestamp.get()
+                val segmentId = currentSegment.get()
+                val replayId = currentReplayId.get()
+
+                val segment =
+                    createSegment(
+                        options.experimental.sessionReplay.sessionSegmentDuration,
+                        currentSegmentTimestamp,
+                        replayId,
+                        segmentId,
+                        height,
+                        width
+                    )
+                if (segment is ReplaySegment.Created) {
+                    segment.capture(hub)
+                    currentSegment.getAndIncrement()
```

Comment:
in case segment creation fails, would you need to increment `currentSegment.getAndIncrement()` as well?

Reply:
no, I think we'd want to try and capture the next time with the same segmentId to avoid gaps. That's a good  point for future though, as we could create an rrweb_event indicating the failure and reflect the gap somehow on the playback. I'll add this to ideas list 👍